src/ui/public/query_bar/directive/query_bar.js

```diff
@@ -15,39 +19,74 @@ module.directive('queryBar', function () {
       query: '=',
       appName: '=?',
       onSubmit: '&',
-      disableAutoFocus: '='
+      disableAutoFocus: '=',
+      indexPattern: '='
     },
     controllerAs: 'queryBar',
     bindToController: true,
-    controller: callAfterBindingsWorkaround(function ($scope, config, PersistedLog) {
+
+    controller: callAfterBindingsWorkaround(function ($scope, $element, $http, $timeout, config, PersistedLog) {
```

Comment:
You're only pulling `$http` in here so you can pass it into the suggestions provider. As noted elsewhere, you don't need it at all, but you could also better encapsulate it by creating a simple kuerySuggestor factory (or service). You could also avoid passing in the config service externally that way as well.

`/src/ui/public/query_bar/service/kuery_suggest.js`
```js
import { uiModules } from 'ui/modules';
import { getSuggestionsProvider } from '../../kuery';

const module = uiModules.get('kibana');

module.factory('kuerySuggestor', function ($http, config) {
  return ({ persistedLog, indexPattern }) => getSuggestionsProvider({ $http, config, persistedLog, indexPattern });
});
```

Then here, you can just add `kuerySuggestor`, and down where you use it, you'd just have `getSuggestions = kuerySuggestor({ indexPattern, persistedLog });`

You could almost certainly take this even further and encapsulate the persisted log part there with a little refactoring of the factory, instead of mixing that concern into the `queryBar` directive here. ie) `getSuggestions.addHistory(this.localQuery.query);`

Further still, you could probably remove the upstream `$scope.$apply` stuff by wrapping `getSuggestionsProvider` in the Angular `Promise` service...

```js
import { uiModules } from 'ui/modules';
import { getSuggestionsProvider } from '../../kuery';

const module = uiModules.get('kibana');

module.factory('KuerySuggestor', function ($http, config, PersistedLog, Promise) {
  return class KuerySuggestor {
    constructor(persistedName, indexPattern) {
      this.persistedLog = new PersistedLog(persistedName, {
        maxLength: config.get('history:limit'),
        filterDuplicates: true
      });
      this.indexPattern = indexPattern;
    }

    getSuggestions() {
      return Promise.resolve(
        getSuggestionsProvider({
          $http,
          config,
          persistedLog: this.persistedLog,
          indexPattern: this.indexPattern
        })
      );
    }

    addHistory(entry) {
      this.persistedLog.add(entry);
    }
  };
});

// controller code:
controller: callAfterBindingsWorkaround(function ($scope, $element, $timeout, config, KuerySuggestor) {

// later in the watcher:
kuerySuggestor = new KuerySuggestor(`typeahead:${this.appName}-${language}`, indexPattern);

// and persisting the query...
this.submit = () => {
  if (this.localQuery.query) kuerySuggestor.add(this.localQuery.query);
  this.onSubmit({ $query: this.localQuery });
};
```

Now the Angularized concerns all live in one place, and the dependencies do too.

Reply:
This would also give you a fine place to decouple the config service and just pass in a config object...

```js
this.suggestionProvider = getSuggestionsProvider({
  $http,
  config: {
    suggestValues: !!config.get('filterEditor:suggestValues'),
  },
  persistedLog: this.persistedLog,
  indexPattern: this.indexPattern
});
```